internal/tools/terraformlinter/terraform_linter.go

```diff
@@ -0,0 +1,335 @@
+type tokenPosition int32
+
+const (
+	None tokenPosition = iota
+	Leading
+	Provider
+	Ignored
+	Trailing
+)
+
+// tokenAttr defines an individual attribute within a block of terraform.
+type tokenAttr struct {
+	tokenPos        tokenPosition
+	token           hclsyntax.Token
+	trailingNewline bool
+}
+
+// keywords to match on.
+const (
+	attrForEach                = "for_each"
+	attrCount                  = "count"
+	attrProvider               = "provider"
+	attrSource                 = "source"
+	attrProviderProject        = "project"
+	attrProviderProjectID      = "project_id"
+	attrProviderFolder         = "folder"
+	attrProviderFolderID       = "foler_id"
+	attrProviderOrganization   = "organization"
+	attrProviderOrganizationID = "organization_id"
+	attrProviderOrgID          = "org_id"
+	attrDependsOn              = "depends_on"
+	attrLifecycle              = "lifecycle"
+)
+
+// mapping of attributes to their expected position.
+var positionMap = map[string]tokenPosition{
+	attrForEach:                Leading,
+	attrCount:                  Leading,
+	attrProvider:               Leading,
+	attrSource:                 Leading,
+	attrProviderProject:        Provider,
+	attrProviderProjectID:      Provider,
+	attrProviderFolder:         Provider,
+	attrProviderFolderID:       Provider,
+	attrProviderOrganization:   Provider,
+	attrProviderOrganizationID: Provider,
+	attrProviderOrgID:          Provider,
+	attrDependsOn:              Trailing,
+	attrLifecycle:              Trailing,
+}
+
+const (
+	errorLeadingMetaBlockAttribute  = "The attribute %q must be in the meta block at the top of the definition."
+	errorMetaBlockNewline           = "The meta block must have an additional new line separating it from the next section."
+	errorProviderAttributes         = "The attribute %q must me below any meta attributes (for_each, count, etc.) but above all other attributes."
+	errorProviderNewline            = "The provider specific attributes must have an additional new line separating it from the next section."
+	errorTrailingMetaBlockAttribute = `The attribute %q must be at the bottom of the resource definition and in the order "depends_on" then "lifecycle."`
+)
+
+// ViolationInstance is an object that contains a reference to a location
+// in a file where a lint violation was detected.
+type ViolationInstance struct {
+	ViolationType string
+	Path          string
+	Line          int
+}
+
+// RunLinter executes the specified linter for a set of files.
+func RunLinter(ctx context.Context, paths []string) error {
+	var violations []*ViolationInstance
+	// Process each provided path looking for violations
+	for _, path := range paths {
+		instances, err := lint(path)
+		if err != nil {
+			return fmt.Errorf("error linting files: %w", err)
+		}
+		violations = append(violations, instances...)
+	}
+	for _, instance := range violations {
+		fmt.Printf("%q detected at [%s:%d]\n", instance.ViolationType, instance.Path, instance.Line)
```

Comment:
Should this be more structured so we can use `reviewdog` or a similar tool to do GitHub-level annotations? I don't think the linter should write to GitHub Actions directly.